[fli-iam#1864] Add NIRS type BIDS

youennmerel opened the pull request:
NIRS BIDS files can now be imported into Shanoir

Specs : https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/11-near-infrared-spectroscopy.html

How to test :
- Import NIRS modality BIDS into Shanoir through the BIDS import
- Check the created examinations/acquisitions/datasets

Data samples : 
https://osf.io/juxy8
https://github.com/bids-standard/bids-examples#nirs-datasets

michaelkain commented on shanoir-ng-front/src/app/import/bids/bids-upload.component.html:
```diff
@@ -71,7 +71,7 @@
 
-<upload-file (fileChange)="uploadArchive($event)" [loading]="archiveStatus == 'uploading'" [error]="archiveStatus == 'error'"></upload-file>
+<upload-file [disabled]="!center" (fileChange)="uploadArchive($event)" [loading]="archiveStatus == 'uploading'" [error]="archiveStatus == 'error'"></upload-file>
 <label *ngIf="archiveStatus == 'uploaded'" > Data sucessufully imported.</label>
```
Hi Youenn, thx! Please correct typo. Thank you!

michaelkain commented on shanoir-ng-front/src/app/datasets/shared/dataset-type.model.ts:
```diff
@@ -31,9 +31,10 @@ export enum DatasetType {
   Statistical = 'Statistical',
   Template = 'Template',
   BIDS = 'BIDS',
+  NIRS = 'NIRS',
```
Hi Youenn, I am wondering about this. From the model point of view all is BIDS: MR, NIRS, EEG, iEEG, etc. Why does NIRS stand here next to BIDS?

youennmerel replied:
You're right, this update is useless.

michaelkain commented:
Hi, I think we could extend the import description to add NIRS [anat/func/dwi/eeg/ieeg/pet/..] on the web page to show, what modality we support.

youennmerel commented on shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/security/DatasetSecurityService.java:
```diff
@@ -574,36 +574,70 @@ public boolean hasUpdateRightOnDataset(Dataset dataset, String rightStr) throws
+	 * @return true or false
+	 * @throws EntityNotFoundException
+	 */
+	public boolean hasUpdateRightOnDatasetAcquisition(DatasetAcquisition datasetAcq, String rightStr) throws EntityNotFoundException {
```
This is a very good question. I've worked on these because this was not working, but I didn't really understood the logic here.

michaelkain replied:
Hi Youenn, it would be great to explore on this please? When the studyId was different (source of the problem), there might be another issue beforehand during the creation of the exam with the frontend (see https://github.com/fli-iam/shanoir-ng/blob/b30c71fd3e8e831f2caee992e1492e68b487fde5/shanoir-ng-import/src/main/java/org/shanoir/ng/importer/bids/BidsImporterApiController.java#L106). I think your change solved the problem, but probably not the origin. Normally, when the exam is correctly created, the studyId must be the same. The user should only be able to create the exam in his studies. Maybe something is wrong with the rabbitmq message to ms datasets to create the correct exam?

youennmerel replied:
Hi @michaelkain,
You're right, checking that studyId did not change make no sense here, as it can't happen when updating an acquisition.

Shouldn't we just test rights on examination, as for an acquisition creation ?

youennmerel replied:
Done.

youennmerel commented:
@michaelkain added 'Nirs' as BIDS data type in study card
![image](https://github.com/fli-iam/shanoir-ng/assets/59697998/f0d2528e-8ed6-4fbc-bb72-8d3a5aaaa0ab)

michaelkain commented:
Very good! Concering my comment on the equipment: I am right, that it remains empty for NIRS in the ds_acquisition? What is by the way totally fine for me.

youennmerel commented:
Added manufacturer model modality type 'Nirs'

youennmerel commented:
> Very good! Concering my comment on the equipment: I am right, that it remains empty for NIRS in the ds_acquisition? What is by the way totally fine for me. I just wanted to avoid dangling "unknown" equipments.

It will remain empty if none are found. However we now can also create 'Nirs' equipment in Shanoir.

youennmerel commented:
Hi @michaelkain,

Should be OK now !

michaelkain commented:
Hi Youenn, thank you for your corrections. I downloaded one of the big NIRS examples above and did some tests, that showed some strange outputs:
1. Good point: the import finishes with success: OK
2. The import message in jobs has an error: [**null** (n°1)] Successfully created datasets for subject [DemoStudy_06] in examination [1] : the study name should not be null in the jobs message
3. Solr download, as BIDS dataset: my subject folder sub-06 contains **34 files** in my original BIDS structure: when I download all datasets imported, I have only **5 files**, what should not be the case. The question is, how do we consider the datasets generated from BIDS. And when I download I have a full failures.txt and the file names in my zip file contain null_null values.
4. the produced BIDS tree looks quite correct (study details), but contains a sub-2-DemoStudy06_scans.tsv  file in Shanoir with 35 lines, but the original contains 36 lines on my disk (to correct)
5. how do you map the files into the datasets in Shanoir? do you interpret the scans.tsv file, that contains 36 fnirs files?
6. the other files in the tree are complete, and have the same naming

michaelkain commented:
Hi, sorry, there is another issue:
![Screenshot from 2023-09-29 11-15-24](https://github.com/fli-iam/shanoir-ng/assets/4560288/7161a98c-d95b-42c6-aa6d-ebdd2fb6071c)
the file names are wrong: they contain two times the subject

michaelkain commented:
more issues:
- the link to the created examination in shanoir is broken, from the created datasets
- same for the created dataset acquisition
- and no equipment is linked, neither created from the information of the BIDS structure

michaelkain reviewed (changes requested):
Hi @youennmerel, please see my comments. Thank you!

youennmerel commented:
Hi @michaelkain,

I took your comments into account, except for (as discussed in daily) :
- Solr BIDS download : should be dealt with in another issue
- scans.tsv content : in your example, the imported scans.tsv is incorrect, the generated one is.

michaelkain commented:
Hi @youennmerel, thank you! The 2 issues above are now fixed. I found two new ones, see my screenshots. I think with multiple sessions you just have to remove the trailing "-" from "ses-1-". I am wondering, as we have the acquisition time in the scans.tsv, if we could not use this as exam date, without time stamp maybe?
![Screenshot from 2023-11-21 17-15-19](https://github.com/fli-iam/shanoir-ng/assets/4560288/41eee82d-d950-4e89-a0b9-541525fd37ef)

michaelkain commented:
and it is strange, that in the Shanoir tree, the dataset do not have the same name "subject name" contained as in the BIDS tree, but maybe this is normal...?
![Screenshot from 2023-11-21 17-23-03](https://github.com/fli-iam/shanoir-ng/assets/4560288/c9673ac4-1f83-42b7-a9fe-1245f19fa19f)
and I think the subject_name BIDS-test06 should be the same everywhere?

michaelkain commented:
Hi @youennmerel, I hope you had a good start into your day. For the <label>s, see BIDS spec:
label - An **alphanumeric value**, possibly prefixed with arbitrary number of 0s for consistent indentation, for example, it is rest in task-rest following task-<label> specification. Note that labels MUST not collide when casing is ignored (see [Case collision intolerance](https://bids-specification.readthedocs.io/en/stable/common-principles.html#case-collision-intolerance)).

youennmerel commented:
Hi @michaelkain,

- Corrected label : now remove non-alphanumeric characters in BIDS tree
- Corrected acq_time parsing : now create examination with acq_time date

michaelkain commented on shanoir-ng-datasets/src/main/java/org/shanoir/ng/bids/service/BIDSServiceImpl.java:
```diff
@@ -775,4 +763,8 @@ private void participantsSerializer(File parentFolder, List<Subject> subjs) {
 		}
 	}
 
+	private String formatLabel(String label){
```
nice move to put this into a dedicated method

michaelkain reviewed (approved):
Nice job @youennmerel !!! Much better looking BIDS tree now, thank you!!!
![Screenshot from 2023-11-24 10-58-45](https://github.com/fli-iam/shanoir-ng/assets/4560288/fadf1db1-63b0-489f-988b-0ac211cdf3db)
